lib/spack/spack/compilers/__init__.py

```diff
@@ -368,6 +412,196 @@ def all_compiler_types():
     return [class_for_compiler_name(c) for c in supported_compilers()]
 
 
+#: Gathers the attribute values by which a detected compiler is considered
+#: unique in Spack.
+#:
+#:  - os: the operating system
+#:  - compiler_name: the name of the compiler (e.g. 'gcc', 'clang', etc.)
+#:  - version: the version of the compiler
+#:
+CompilerID = collections.namedtuple(
+    'CompilerID', ['os', 'compiler_name', 'version']
+)
+
+#: Variations on a matched compiler name
+NameVariation = collections.namedtuple('NameVariation', ['prefix', 'suffix'])
+
+#: Groups together the arguments needed by `detect_version`. The four entries
+#: in the tuple are:
+#:
+#: - id: An instance of the CompilerID named tuple (version can be set to None
+#:       as it will be detected later)
+#: - variation: a NameVariation for file being tested
+#: - language: compiler language being tested (one of 'cc', 'cxx', 'fc', 'f77')
+#: - path: full path to the executable being tested
+#:
+DetectVersionArgs = collections.namedtuple(
+    'DetectVersionArgs', ['id', 'variation', 'language', 'path']
+)
+
+
+def arguments_to_detect_version_fn(operating_system, paths, override=True):
```

Comment:
@tgamblin @becker33 I moved all the functions that are needed to search for compilers in this module, as I think they belong here as free functions and not in `spack.architecture.OperatingSystem` as methods.  I had to leave room for overriding when porting Cray compilers, as that's currently the only architecture that needs a special treatment. 

I am not particularly proud of the `override=True` in the arguments above, that is needed to re-use the `_default` implementation for Cray front-end, and will probably spend some time looking for alternative approaches that leave the default behavior as a nested function that is not visible externally. Any idea is welcome.

Reply:
@alalazo: I don't like it either -- if you can figure out how best to get rid of it that would be awesome.

Reply:
I had a better look at the code and the part where I need to override the function was basically to compute different search paths. Therefore I added another point of customization for the OS in aa702a0. The result looks cleaner to me than before. 

If the PR is fine with you in the current state I'll write some docs on how to add a new compiler and ping you back. Let me know if I should proceed with that.